programs/staking/src/lib.rs

```diff
@@ -11,46 +11,74 @@ declare_id!("CBbYHhjfhFoPwz8ZgQJjHkpq2gSu1xn26qeVsfYhWWB5");
 mod staking {
     use super::*;
 
-    pub fn initialize_vault(ctx: Context<InitializeVault>, bumps: VaultBumps) -> Result<()> {
+    pub fn initialize_vault(ctx: Context<InitializeVault>, vault_bump: u8) -> Result<()> {
         let vault = &mut ctx.accounts.vault;
-
-        vault.bumps = bumps;
-        vault.vault_mer = ctx.accounts.vault_mer.key();
+        vault.vault_bump = vault_bump;
+        vault.token_mint = ctx.accounts.token_mint.key();
+        vault.token_vault = ctx.accounts.token_vault.key();
         vault.lp_mint = ctx.accounts.lp_mint.key();
+        vault.base  = *ctx.accounts.base.key;
         vault.admin = *ctx.accounts.admin.key;
+        vault.locked_reward_tracker = LockedRewardTracker::default();
+        Ok(())
+    }
+
+    // transfer admin. Ex: to gorvernence
+    pub fn transfer_admin(ctx: Context<TransferAdmin>) -> Result<()> {
+        let vault = &mut ctx.accounts.vault;
+        vault.admin = *ctx.accounts.new_admin.key;
+        Ok(())
+    }
 
+    pub fn update_locked_reward_degradation(
+        ctx: Context<UpdateLockedRewardDegradation>,
+        locked_reward_degradation: u64,
+    ) -> Result<()> {
+        let vault = &mut ctx.accounts.vault;
+        if locked_reward_degradation > u64::try_from(LOCKED_REWARD_DEGRATION_DENUMERATOR).unwrap() {
```

Comment:
Why don't upscale `locked_reward_degradation` which is safe, instead of down scaling `u128` to `u64` and unwrap?

Reply:
LOCKED_REWARD_DEGRATION_DENUMERATOR is a fixed number, so it is safe to be casted to u64. I want to keep it as u128 to save computation for casting when calculating locked reward